### ISSUE TITLE
deepcopy and caching lead to very strange things

### ISSUE THREAD

**Issue by pbrady** [sympy/sympy#7672]
Ran into some testing errors in the solvers module when working on #7464 that seemed to be related to deepcopy.  I was able to reproduce this using the current caching mechanism in Sympy:

```
>>> from sympy import symbols
>>> from copy import deepcopy
>>> x, xx = symbols('x'), symbols('x', real=True)
>>> x.assumptions0
{'commutative': True}
>>> xx.assumptions0
{'real': True, 'imaginary': False, 'complex': True, 'hermitian': True, 'commutative': True}
>>> x == xx
False
>>> xxx = deepcopy(xx)
>>> x == xx
True
>>> x.assumptions0
{'real': True, 'imaginary': False, 'complex': True, 'hermitian': True, 'commutative': True}
```

This works as expected with SYMPY_USE_CACHE=no.  Seems that deepcopy is grabbing an item (the wrong one) from the cache and modifying it.  Weird.

**Comment by dvarjas** [comment]
I experience exactly the same problem using `copy`. Here is a short code segment that shows, by copying it is possible to have two symbols that look identical (`c0` and `c0_new`), but hash different, and one of them (`c0`) has a cached hash that is inconsistent with the recalculated hash.
import sympy
from copy import copy

c0 = sympy.symbols('c0')
print('c0:')
print(c0, c0.assumptions0)
h = hash(c0)
h2 = hash((type(c0).__name__,) + c0._hashable_content())
print(h == h2, h, h2)
print('----')

c0_new = sympy.symbols('c0', real=True)
print('c0_new:')
print(c0_new, c0_new.assumptions0)
h = hash(c0_new)
h2 = hash((type(c0_new).__name__,) + c0_new._hashable_content())
print(h == h2, h, h2)
print('c0 is c0_new', c0 is c0_new)
print('----')

c = copy(c0_new)
print('c:')
print(c, c.assumptions0)
h = hash(c)
h2 = hash((type(c).__name__,) + c._hashable_content())
print(h == h2, h, h2
      # hash(c.name), hash(tuple(sorted(c.assumptions0.items()))), hash(type(c)), c._mhash,
      )
print('c is c0_new', c is c0_new)
print('c is c0', c is c0)

print('----')
print('c0:')
print(c0, c0.assumptions0)
h = hash(c0)
h2 = hash((type(c0).__name__,) + c0._hashable_content())
print(h == h2, h, h2)
print('c0 is c0_new', c0 is c0_new)

print('----')
print('c0_new:')
print(c0_new, c0_new.assumptions0)
h = hash(c0_new)
h2 = hash((type(c0_new).__name__,) + c0_new._hashable_content())
print(h == h2, h, h2)
print('c0 is c0_new', c0 is c0_new)
print('----')

**Comment by oscarbenjamin** [comment]
This is explained above: https://github.com/sympy/sympy/issues/7672#issuecomment-47578123

See the docs for pickle:
https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-normal-class-instances
and the methods in Symbol:
https://github.com/sympy/sympy/blob/4b93506a88cb924785a0e1dc98bb15205d2d45b8/sympy/core/symbol.py#L266-L270

When you copy an object it is effectively pickled and unpickled again. The unpickling first creates a Symbol object with `Symbol('x')` and then sets the `_assumptions` attribute afterwards. Calling `Symbol('x')` will retrieve the cached symbol. Afterwards it is modified by `__setstate__`.

One way to fix this would be to make it so that Symbol can accept an additional positional argument representing the assumptions somehow. Then `__getnewargs__` could return that. There are a few bits of hacky business in Symbol that could maybe be cleaned up by doing that.

**Comment by dvarjas** [comment]
I see, the issue here is that `__getnewargs__` doesn't support `kwargs`. I found [this hacky way](__getnewargs__) around it, but it will likely only make the code even messier.

A somewhat cleaner alternative could be to define `__setstate__` and make it recalculate the hash with the full final state.

**Comment by oscarbenjamin** [comment]
Using `__setstate__` will not help. It will still be modifying the object retrieved from the cache that will be used by other code that didn't expect the assumptions to be modified.

The clean way is to make it possible to create a Symbol with assumptions without using kwargs.

**Comment by oscarbenjamin** [comment]
This could be fixed easily defining `__getnewargs_ex__` for Symbol. We would need to remove the `__reduce__` method defined on `Basic` though:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index 8e82778c7d..798c7fbc0c 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -121,11 +121,7 @@ def __new__(cls, *args):
     def copy(self):
         return self.func(*self.args)
 
-    def __reduce_ex__(self, proto):
-        """ Pickling support."""
-        return type(self), self.__getnewargs__(), self.__getstate__()
-
     def __getnewargs__(self):
        return self.args
 
     def __getstate__(self):
diff --git a/sympy/core/symbol.py b/sympy/core/symbol.py
index 68a5d000f1..90dd835299 100644
--- a/sympy/core/symbol.py
+++ b/sympy/core/symbol.py
@@ -300,11 +300,8 @@ def __new_stage2__(cls, name, **assumptions):
     __xnew_cached_ = staticmethod(
         cacheit(__new_stage2__))   # symbols are always cached
 
-    def __getnewargs__(self):
-        return (self.name,)
-
-    def __getstate__(self):
-        return {'_assumptions': self._assumptions}
+    def __getnewargs_ex__(self):
+        return ((self.name,), self.assumptions0)
 
     def _hashable_content(self):
         # Note: user-specified assumptions not hashed, just derived ones
```

**Comment by eagleoflqj** [comment]
Should be fixed in https://github.com/sympy/sympy/pull/21260. Could close? @oscarbenjamin

**Comment by oscarbenjamin** [comment]
Yes, thanks for pointing this out. It was fixed by #21260
